cache.go

```diff
@@ -208,6 +212,109 @@ func NewCache(config *Config) (*Cache, error) {
 	return cache, nil
 }
 
+// Snapshot creates a snapshot of the state of the cache
+// Note: we do this with minimal locking for performance reasons. Namely
+// we do not attempt to form a proper point-in-time view of the LFU structures,
+// metrics and sharded map data, preferring a "best effort" export of the data and
+// allowing for some drift in the admission/eviction policy structures and metrics.
+func (c *Cache) Snapshot(dir string) error {
+	var err error
+	var dirInfo os.FileInfo
+
+	// check that dir exists and is writable
+	if dirInfo, err = os.Stat(dir); os.IsNotExist(err) {
+		return err
+	}
+	if !dirInfo.IsDir() {
+		return errors.New("snapshot path is not a directory")
+	}
+
+	// snapshot policy
+	err = c.policy.Snapshot(dir)
+	if err != nil {
+		return err
+	}
+
+	// snapshot cache metrics
+	err = c.Metrics.Snapshot(dir)
+	if err != nil {
+		return err
+	}
+
+	return c.store.Snapshot(dir)
+}
```

Comment:
Since we're using mmap'd buffers I thought it'd be better to open them individually and keep them open only as long as needed. Curious what race conditions you're worried about? I don't forsee any but I may be missing something.

Reply:
We can still open the buffers one at a time, I'm suggesting to open handles to the files at the beginning.

The race conditions (I've realized this wasn't clear) are scenarios where we can't open all the files or if something happens to files the snapshotter has already created. In other words, it's better to test and acquire a "lock" on the files we're going to write to in the beginning.

Reply:
Oh I see - to avoid any extraneous work if we fail to create some files?

Reply:
Yeah, that's one failure scenario.

Reply:
I also see your concern with a partial snapshot... which as you say can happen for a number of reasons. In this case we'll return an error from the ristretto call and I guess assume the client will do something about it (in terms of recognizing the bad state on the filesystem, cleaning up the files in question, etc). I think this is a reasonable assumption from the ristretto side but also open to moving some of that handling here.

Reply:
Yeah, to be fair, it's not a requirement to merging this in.